mmhuman3d/core/visualization/renderer/torch3d_renderer/uv_renderer.py

```diff
@@ -0,0 +1,537 @@
+        assert displacement.shape[1] == self.NUM_VERTS
+        assert displacement.shape[0] in [batch_size, 1]
+
+        if displacement.shape[0] == 1:
+            displacement = displacement.repeat(batch_size, 1, 1)
+        C = displacement.shape[-1]
+        if C == 1:
+            displacement = meshes.verts_normals_padded() * displacement
+
+        displacement = padded_to_packed(displacement)
+
+        meshes = meshes.to(self.device)
+        meshes = meshes.offset_verts(displacement)
+        return meshes
+
+    def wrap_texture(self,
+                     meshes: Meshes,
+                     texture_map: torch.Tensor,
+                     resolution: Optional[Iterable[int]] = None,
+                     mode: Optional[str] = 'bicubic',
+                     is_bgr: bool = True) -> Meshes:
+        """Wrap a texture map to the input meshes.
+
+        Args:
+            meshes (Meshes): the input meshes.
+            texture_map (torch.Tensor): the texture map to be wrapped.
+                Shape should be (N, H, W, 3)
+            resolution (Optional[Iterable[int]], optional): resolution to
+                override self.resolution. If None, will use self.resolution.
+                Defaults to None.
+            mode (Optional[str], optional): interpolate mode.
+                Should be in ['nearest', 'bilinear', 'trilinear', 'bicubic',
+                'area'].
+                Defaults to 'bicubic'.
+            is_bgr (bool, optional): Whether the color channel is BGR.
+                Defaults to True.
+
+        Returns:
+            Meshes: returned meshes.
+        """
+
+        batch_size = len(meshes)
+        if texture_map.ndim == 3:
+            texture_map_padded = texture_map[None]
+        else:
+            texture_map_padded = texture_map
+        assert texture_map.shape[-1] == 3
+        _, H, W, _ = texture_map_padded.shape
+
+        resolution = resolution if resolution is not None else (H, W)
+
+        if resolution != (H, W):
+            texture_map_padded = F.interpolate(
+                texture_map_padded.view(0, 3, 1, 2), resolution,
+                mode=mode).view(0, 2, 3, 1)
+        assert texture_map_padded.shape[0] in [batch_size, 1]
+
+        if isinstance(texture_map_padded, np.ndarray):
+            texture_map_padded = self.array2tensor(texture_map_padded)
+            is_bgr = True
+        if is_bgr:
+            texture_map_padded = self.rgb2bgr(texture_map_padded)
+
+        if texture_map_padded.shape[0] == 1:
+            texture_map_padded = texture_map_padded.repeat(batch_size, 1, 1, 1)
+
+        faces_uvs = self.faces_uv[None].repeat(batch_size, 1, 1)
+        verts_uvs = self.verts_uv[None].repeat(batch_size, 1, 1)
+        textures = TexturesUV(
+            faces_uvs=faces_uvs, verts_uvs=verts_uvs, maps=texture_map_padded)
+        meshes.textures = textures
+        return meshes
+
+    @staticmethod
+    def rgb2bgr(rgbs) -> Union[torch.Tensor, np.ndarray]:
```

Comment:
It has been defined in MeshBaseRenderer. Could be reused.

Reply:
> It has been defined in MeshBaseRenderer. Could be reused.

ok